internal/parser/grammar.md

```diff
@@ -0,0 +1,71 @@
+    VariableDeclaration = ident assign Expression .
+    ReturnStatement = return Expression .
+    ExpressionStatement = Expression .
+    Expression = LogicalExpression .
+    ExpressionList = [ Expression { comma Expression } ] .
+    PropertyList = [ Property { comma Property } ] .
+    Property = ident colon Expression .
+    ParameterList = [ Parameter { comma Parameter } ] .
+    Parameter = ident eq Expression .
+    LogicalExpression = ComparisonExpression
+                      | LogicalExpression LogicalOperator ComparisonExpression .
+    LogicalOperator = and | or .
+    ComparisonExpression = MultiplicativeExpression
+                         | ComparisonExpression ComparisonOperator MultiplicativeExpression .
+    ComparisonOperator = eq | neq | regexeq | regexneq .
+    MultiplicativeExpression = AdditiveExpression
+                             | MultiplicativeExpression MultiplicativeOperator AdditiveExpression .
+    MultiplicativeOperator = mul | div .
+    AdditiveExpression = PipeExpression
+                       | AdditiveExpression AdditiveOperator PipeExpression .
+    AdditiveOperator = add | sub .
+    PipeExpression = PostfixExpression
+                   | PipeExpression PipeOperator PostfixExpression .
+    PipeOperator = pipe_forward .
+    PostfixExpression = UnaryExpression
+                      | PostfixExpression PostfixOperator .
+    PostfixOperator = DotExpression
+                    | CallExpression
+                    | IndexExpression .
+    DotExpression = dot ident .
+    CallExpression = lparen ParameterList rparen .
+    IndexExpression = lbrack Expression rbrack .
+    UnaryExpression = PrimaryExpression
+                    | PrefixOperator PrimaryExpression .
+    PrefixOperator = add | sub | not .
+    PrimaryExpression = ident
+                      | int
+                      | float
+                      | string
+                      | regex
+                      | duration
+                      | pipe_receive
+                      | ObjectLiteral
+                      | ArrayLiteral
+                      | ParenExpression
+                      | ArrowExpression .
+    ObjectLiteral = lbrace PropertyList rbrace .
+    ArrayLiteral = lbrack ExpressionList rbrack .
+    ParenExpression = lparen Expression rparen .
+    ArrowExpression = lparen ParameterList rparen arrow ArrowBodyExpression .
+    ArrowBodyExpression = BlockStatement | Expression .
+    BlockStatement = lbrace StatementList rbrace .
+
+Note: The "option" is an identifier with the literal "option". It will have the `ident` token type, but must contain that exact literal to follow that branch. It is possible for the "option" identifier to also be interpreted as a normal identifier as it is not a reserved keyword.
+
+For the parser, the above grammar undergoes a process to have the left-recursion removed and is then left-factored to turn it into an LL(1) compliant grammar.
+
+    TODO
```

Comment:
This is what I had in mind, that way it is easy to compare the natural EBNF to the refactored one and then compare the refactored one to the code.